### PR TITLE
feat: app test

### PR DESCRIPTION
**Pull request by rjzak**
Working on signed tag and directory creation.

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

File: crates/app/tests/mod.rs [hunk 1]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+    println!("Creating signed tag.");
+    tag::create(
+        &cl,
+        &addr,
+        &foo,
+        &tag_2,
+        Jws::TYPE,
+        TagEntry::Signed(Jws::Flattened(Flattened {
+            payload: Some(Bytes::from(random_bytes_sha256)),
+            signature: Signature {
```
Comment (rjzak) [inline crates/app/tests/mod.rs:132]: These fields in the Signed tag are causing stack overflow errors. Am I missing something obvious?

File: crates/app/tests/mod.rs [hunk 2]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
 
+    println!("Put testing file in foo tag");
+
     let (test_resp, test_type) =
         tree::get_path(&cl, &addr, &foo, &tag, &"/".parse().unwrap()).await;
 
     assert_eq!(test_type, mime::TEXT_PLAIN);
     assert_eq!(&test_resp[..], b"testing");
 
+    let random_bytes = [
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:106]: This is unmaintainable, what is this variable for? Where did these values come from?
Apart from that `u8` is redundant

If you need random bytes, generate them at runtime.
If you need a constant, use a human-readable string `const`.

File: crates/app/tests/mod.rs [hunk 3]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+        .to_vec();
+
+    let tag_2 = "0.2.0".parse().unwrap();
+    println!("Creating signed tag.");
+    tag::create(
+        &cl,
+        &addr,
+        &foo,
+        &tag_2,
+        Jws::TYPE,
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:129]: This means you're creating a signed tag of a signed tag

File: crates/app/tests/mod.rs [hunk 4]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+        TagEntry::Signed(Jws::Flattened(Flattened {
+            payload: Some(Bytes::from(random_bytes_sha256)),
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:131]: Why not just store a constant `"test"`? What's the benefit of using a hash here?

File: crates/app/tests/mod.rs [hunk 5]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+            signature: Signature {
+                protected: None,
+                header: None,
+                signature: Default::default(),
+            },
+        })),
+    )
+    .await;*/
+
+    println!("Created tag2");
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:163]: This is too verbose to be printed on each test execution, as a rule of thumb, you should only see output if something is broken

File: crates/app/tests/mod.rs [hunk 6]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+
+    let two_tags = tag::list(&cl, &addr, &foo).await;
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:165]:
```suggestion
    let tags = tag::list(&cl, &addr, &foo).await;
```

File: crates/app/tests/mod.rs [hunk 7]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+    assert!(two_tags.contains(&tag));
+    assert!(two_tags.contains(&tag_2));
+
+    assert_eq!(tag::list(&cl, &addr, &bar).await, vec![]);
+
+    println!("Get two tags assertions passed.");
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:171]: see above, there's no value for this statement, if the assertion wouldn't have passed - you'd see a failure

File: crates/app/tests/mod.rs [hunk 8]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+
+    tree::create_path(
+        &cl,
+        &addr,
+        &foo,
+        &tag_2,
+        &"/".parse().unwrap(),
+        Directory::TYPE.parse().unwrap(),
+        b"/sub_directory".to_vec(),
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:180]: This should be a JSON-encoded object mapping names to entries.

File: crates/app/tests/mod.rs [hunk 9]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+    )
+    .await;
+    println!("Created directory");
+
+    tree::create_path(
+        &cl,
+        &addr,
+        &foo,
+        &tag_2,
+        &"/sub_directory/testing_file.bin".parse().unwrap(),
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:190]: this is very verbose, while the verbosity does not bring any value
```suggestion
        &"/foo/bar".parse().unwrap(),
```

File: crates/app/tests/mod.rs [hunk 10]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+        mime::APPLICATION_OCTET_STREAM,
+        random_bytes.to_vec(),
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:192]: Why not `b"test".to_vec()`?

File: crates/app/tests/mod.rs [hunk 11]
```diff
@@ -89,12 +95,117 @@ async fn app() {
...
+    )
+    .await;
+    println!("Created testing file in dir");
+
+    let (test_resp, test_type) = tree::get_path(
+        &cl,
+        &addr,
+        &foo,
+        &tag_2,
+        &"/sub_directory/testing_file.bin".parse().unwrap(),
```
Comment (rvolosatovs) [inline crates/app/tests/mod.rs:202]: see above

File: crates/jose/src/jws.rs [hunk 12]
```diff
@@ -45,6 +45,10 @@ impl MediaTyped for Jws {
     const TYPE: &'static str = "application/jose+json";
 }
 
+impl Jws {
+    pub const TYPE: &'static str = "application/jose+json";
```
Comment (rvolosatovs) [inline crates/jose/src/jws.rs:49]: Redundant, see https://github.com/profianinc/drawbridge/pull/141/files#diff-f489372ea9f1f61671d0178d79df7eb4bd7d9ebcf7b0d7e8308592705e2c5732R45 right above

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Comment by rvolosatovs** [comment]
Please work on your commit messages.
See https://cbea.ms/git-commit/ and then https://www.conventionalcommits.org/en/v1.0.0/

File: crates/app/tests/mod.rs (hunk 1 above)
Comment (rvolosatovs) [reply crates/app/tests/mod.rs:132]: The `random_bytes_sha256` is not a valid `JWS` (you're trying to store a signed tag in a signed tag)

Comment (rjzak) [reply crates/app/tests/mod.rs:132]: How do I use JWS properly? Removing anything and the compiler complains about incorrect datatype.

**Pull request closed without merging**